File: 852.peak-index-in-a-mountain-array.java
/*
 * @lc app=leetcode id=852 lang=java
 *
 * [852] Peak Index in a Mountain Array
 */

// @lc code=start
class Solution {
    public int peakIndexInMountainArray(int[] arr) {
        int left = 0, right = arr.length - 1;
        while (left < right) {
            int pivot = left + (right - left) / 2;
            if (arr[pivot] > arr[pivot + 1]) {
                right = pivot;
            } else {
                left = pivot + 1;
            }
        }
        return left;
        // throw new IllegalArgumentException("No Solution");
    }
}
// @lc code=end

File: 852.peak-index-in-a-mountain-array.py
#
# @lc app=leetcode id=852 lang=python3
#
# [852] Peak Index in a Mountain Array
#

# @lc code=start

class Solution:
    def peakIndexInMountainArray(self, arr: List[int]) -> int:
        # guaranteed to be a mountain
        left, right = 0, len(arr)-1
        while left < right:
            mid = (left+right)//2
            # of course, arr[mid+1] exists
            if arr[mid] < arr[mid+1]:
                left = mid+1
            else:
                right = mid
        return left


# @lc code=end
